src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -438,6 +440,20 @@ private static GeneratorState SetGeneratorException(CommonMessageProvider provid
             return new GeneratorState(generatorState.Info, e, diagnostic);
         }
 
+        private static ImmutableArray<Diagnostic> FilterDiagnostics(Compilation compilation, ImmutableArray<Diagnostic> diagnostics, CancellationToken cancellationToken)
+        {
+            DiagnosticBag filteredDiagnostics = DiagnosticBag.GetInstance();
```

Comment:
A `DiagnosticBag` instance is somewhat heavy weight because it's designed for concurrent writes. In this case we're just building up a collection in a single threaded environment and returning it to the caller. Consider using a normal `ImmutableArrayBuilder` here. 

Given that this is immediately consumed and written into a different bag there isn't really a need for the return to be immutable here if that adds extra cost.

Reply:
Thanks. Changed to `ArrayBuilder` but also passing in the overal driver bag, so we can just add them as we go rather than adding them as a bunch at the end.